### PR TITLE
Implement Pow(12) for cpu and cuda

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**: 
Implement Pow(12) supporting int32, int64, float, double for both base and exponent.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/math/element_wise_ops.cc [hunk 1]
```diff
@@ -273,25 +275,89 @@ Status Sqrt<T>::Compute(OpKernelContext* ctx) const {
...
+      break;
+    case on::TensorProto_DataType_FLOAT:
+      PowImpl<B, float>(context, X, Y);
+      break;
+    case on::TensorProto_DataType_DOUBLE:
+      PowImpl<B, double>(context, X, Y);
+      break;
+    default:
+      assert(false);
   }
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/math/element_wise_ops.cc:330]: Can this return Status with an error message saying an unexpected type was found?

File: onnxruntime/core/providers/cpu/math/element_wise_ops.cc [hunk 2]
```diff
@@ -273,25 +275,89 @@ Status Sqrt<T>::Compute(OpKernelContext* ctx) const {
...
-  const Tensor& Y = *context->Input<Tensor>(1);
-  std::function<void(EigenVectorMap<T>, ConstEigenVectorMap<T>, T)> input1scalar =
-      [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T input1) { output = Eigen::pow(input0.array(), input1); };
-  if (Y.Shape().Size() == 1) {
-    T value = *Y.Data<T>();
-    if (value == 2.0) {
-      input1scalar = [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T) { output = Eigen::square(input0.array()); };
-    } else if (value == 3.0) {
-      input1scalar = [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T) { output = Eigen::cube(input0.array()); };
-    }
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/math/element_wise_ops.cc:287]: Eigen::square and Eigen::cube are massively faster with floationg point values  than Eigen::pow. If that also applies to std::pow we need to keep these special cases.

File: onnxruntime/core/providers/cuda/cu_inc/common.cuh [hunk 3]
```diff
@@ -181,8 +181,11 @@ __device__ __inline__ half2 _Tanh(half2 a) {
   return __float22half2_rn(tmp);
 }
 
-template <typename T>
-__device__ __inline__ T _Pow(T a, T b);
+// Capture permutations of int32/64/float/double
+template <typename T, typename T1>
+__device__ __inline__ T _Pow(T a, T1 b) {
+  return static_cast<T>(pow(static_cast<double>(a), static_cast<double>(b)));
+}
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/cu_inc/common.cuh:188]: Any potential accuracy issues with this approach?

File: onnxruntime/test/providers/cpu/math/element_wise_ops_test.cc [hunk 4]
```diff
@@ -621,6 +621,60 @@ TEST(MathOpTest, Pow_Broadcast_Scalar1) {
   test.Run();
 }
 
+TEST(MathOpTest, Pow_float_int64) {
+  OpTester test("Pow", 12);
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/element_wise_ops_test.cc:625]: Can you add some tests with broadcasting? Or run some of the existing tests with both the original opset and opset 12?

File: onnxruntime/core/providers/cuda/cu_inc/common.cuh (hunk 3 above)
Comment (yuslepukhin) [reply onnxruntime/core/providers/cuda/cu_inc/common.cuh:188]: I think int64 may potentially loose precision. However, its support is needed and CUDA only supports pow(double, double) and powf(float, float). No  mixed types which Pow(12) requires.

---
In reply to: [417016905](https://github.com/microsoft/onnxruntime/pull/3727#discussion_r417016905) [](ancestors = 417016905)

Comment (yuslepukhin) [reply onnxruntime/core/providers/cuda/cu_inc/common.cuh:188]: I just confirmed with Pytorch code they do the same thing.

---
In reply to: [417529223](https://github.com/microsoft/onnxruntime/pull/3727#discussion_r417529223) [](ancestors = 417529223,417016905)

**Pull request merged**